### DISCUSSION TITLE
Inject members only once when TestInstance.Lifecycle.PER_CLASS

### DISCUSSION THREAD

**Discussion by harel-biton** [xvik/dropwizard-guicey#394] [Ideas]
Most of my tests (integration) are annotated with `TestInstance.Lifecycle.PER_CLASS`. It works great with my application and feels more natural since the "embedded application" that is being run by the `TestGuiceyAppExtension` is also one per the whole class.
I am in the process of improving the performance of my tests, and I noticed that the `TestGuiceyAppExtension` injects the members of the test instance **before each test method**, regardless of the `TestInstance.Lifecycle` of the tests class ([here](https://github.com/xvik/dropwizard-guicey/blob/3dc6037cbff204a865247d5d931cc88496366fe9/dropwizard-guicey/src/main/java/ru/vyarus/dropwizard/guice/test/jupiter/ext/GuiceyExtensionsSupport.java#L145)).

IMO, it makes sense that when the test is annotated with `TestInstance.Lifecycle.PER_CLASS`, the members will be injected once before all the test methods.
It would be great if I could somehow configure the test extension so it will behave this way.

**Comment by xvik** [comment]
I did it like this for simplicity (there are already a lot of logic):  such "duplicate" injections assumed to take almost no time. Considering that guice context itself is not re-created, the same dependencies would be autowired each time so it's safe.

There is an additional benefit for prototype scoped beans - a new instance of prototype beans would be injected before each  test method (could be handy).

Of course, its not so hard to add an additional option to change the default behavior, but I'm not sure it worth efforts.
What benefits it would bring? Would it lead to a performance gain (considering all tests)? 

I understand that it looks as "not quite correct", but there is no harm and no performance hit (should not be). Most users could even never know about this behaviour. I'm open for discussion.

Just to mention:

>  the "embedded application" that is being run by the TestGuiceyAppExtension is also one per the whole class

Yes, extension annotation behaves like that, but you can declare extension in a non-static filed and it [would start new application for each test method](https://xvik.github.io/dropwizard-guicey/latest/guide/test/junit5/#start-application-by-test-method):

```java
@RegisterExtension
TestGuiceyAppExtension ext = TestGuiceyAppExtension.forApp(App.class).create()
```

**Comment by harel-biton** [comment]
Thanks for the reply.
Yes, my motivation is to improve my test performance, and I do see the members injection affects it.

I just ran all my tests (about ~60 test classes) and measured the time of the "unneeded injections":
I created an extension that wraps `TestGuiceyAppExtension` and delegates all the calls to it. In the `beforeEach` function, I measured the time of the function in case it was not the first time it had been executed for this test class (all my tests are `PER_CLASS`, and the extension is defined as a static field in all of them). 

The total time was almost 43 seconds.

I have to say that I have 50 more test classes using some other framework (not `TestGuiceyAppExtension`), and I plan to move them to `TestGuiceyAppExtension`, so the time might be doubled.

**Comment by xvik** [comment]
Oh, 43 seconds is too much. Then I will add this option (keep current behaviour as default).

If possible, could you please show an example test (or describe the specifics of your tests): I'm just curious why it takes so much time for injection - I can think of too many injection points, but it could also be the amount of nested tests or something else.

**Comment by harel-biton** [comment]
Thanks!

Regarding my tests:
- I don't have any nested tests
- Most of my test classes have at most 10 injections, but they are all extending a base class that contains some more injections. I am also working on trying to minimize this

**Comment by xvik** [comment]
Sorry for delay, upcoming version would have `injectOnce` option (`@TestGuiceyApp(value = App.class, injectOnce = true)`)  disabling duplicate injections for the same test instance.

Also, debug option (`@TestGuiceyApp(value = App.class, debug = true)`)  was improved with performance report, so now you could easilly see how much time spent by guicey extension. Report shown after each `BeforeEach` and `AfterAll` to let you see the time spent for each test. For example, test with `injectOnce` enabled (PerClass):

```
\\\-----------------------------------------------------------------/ test instance = 998458782 /
Guicey time after [Before each] of PerClassInjectOnceGuiceyTest#injectTest1(): 860.7 ms

	[Before all]                       : 858.48 ms
		Guicey fields search               : 1.18 ms
		Guicey hooks registration          : 0.02 ms
		DropwizardTestSupport creation     : 1.78 ms
		Application start                  : 848.08 ms

	[Before each]                      : 2.23 ms
		Guice fields injection             : 1.35 ms


\\\-----------------------------------------------------------------/ test instance = 998458782 /
Guicey time after [Before each] of PerClassInjectOnceGuiceyTest#injectTest2(): 860.82 ms ( + 0.12 ms)

	[Before all]                       : 858.48 ms
		Guicey fields search               : 1.18 ms
		Guicey hooks registration          : 0.02 ms
		DropwizardTestSupport creation     : 1.78 ms
		Application start                  : 848.08 ms

	[Before each]                      : 2.3 ms ( + 0.08 ms)
		Guice fields injection             : 1.35 ms

	[After each]                       : 0.05 ms

```

Two test methods executed for the same test instance. Members were not injected for the second one:
```
		Guice fields injection             : 1.35 ms
```
(no braces with plus means no addition execution)

**Comment by xvik** [comment]
[7.2.0 released](https://xvik.github.io/dropwizard-guicey/7.2.0/about/release-notes/) with injectOnce option, described above and [test performance report](https://xvik.github.io/dropwizard-guicey/7.2.0/guide/test/junit5/debug/#startup-performance) (activated in debug mode).
